IdentityCore/src/util/MSIDKeyOperationUtil.m

```diff
@@ -0,0 +1,129 @@
+        [self generateErrorWithMessage:errorMessage underlyingError:nil context:context error:error];
+    }
+    return isSupported;
+}
+
+- (MSIDJwtAlgorithm)getJwtAlgorithmForKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    if (privateKey == NULL)
+    {
+        [self generateErrorWithMessage:@"No key to determine signing algorithm" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmECDSASignatureDigestX962SHA256 key:privateKey context:context error:error])
+    {
+        return MSID_JWT_ALG_ES256;
+    }
+    
+    if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256 key:privateKey context:context error:error])
+    {
+        return MSID_JWT_ALG_RS256;
+    }
+    
+    if (error)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Key does not support signing any supported algorithms or key is public key");
+    }
+    return nil;
+}
+
+- (NSData *)getSignatureForDataWithKey:(NSData *)rawData privateKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext>)context error:(NSError * _Nullable __autoreleasing *)error
+{
+    if (!rawData)
+    {
+        [self generateErrorWithMessage: @"No data to sign" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    if (privateKey == NULL)
+    {
+        [self generateErrorWithMessage:@"No key defined to sign data with" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    SecKeyAlgorithm algorithm = nil;
+    // Since Secure enclave only supports ECC NIST P-256 curve key we can assume key is used for ECDSA
+    if ([[MSIDKeyOperationUtil sharedInstance] isKeyFromSecureEnclave:privateKey] && [[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmECDSASignatureMessageX962SHA256 key:privateKey context:context error:error])
+    {
+        algorithm = kSecKeyAlgorithmECDSASignatureMessageX962SHA256;
+    }
+    else if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256 key:privateKey context:context error:error])
+    {
+        algorithm = kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256;
+    }
+    
+    if (algorithm == NULL)
+    {
+        [self generateErrorWithMessage:@"Signing algorithm could not be determined for supplied key because key is not of supported type or it is a public key." underlyingError:nil context:context error:error];
+        return nil;
+    }
+    CFErrorRef *subError = NULL;
+    NSData *signature = (NSData *)CFBridgingRelease(SecKeyCreateSignature(privateKey,
```

Comment:
nit: should we first assign it to CFDataRef variable and do bridging Release if and only if it is not nil and there is no error?

Reply:
I assume in this case NSData* signature will be nil if there is any error. The output is nil when trying to CFBridgingRelease(NULL)